File: java/oft_primes/src/test/java/PrimeTest.java
import com.nealford.functionalthinking.primes.Prime;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static com.nealford.functionalthinking.primes.Prime.*;
import static java.util.Collections.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class PrimeTest {
  private ArrayList<Integer> PRIMES_BELOW_50 = new ArrayList<Integer>() {{
              add(2);  add(3);  add(5);  add(7);  add(11);  add(13);
              add(17); add(19); add(23); add(29); add(31);  add(37);
              add(41); add(43); add(47);
          }};

  private Set<Integer> setOf(int... numbers) {
    Set<Integer> result = new HashSet<>();
    for (int i : numbers)
      result.add(i);
    return result;
  }

  @Test
  public void factors() {
    assertEquals(setOf(1, 2, 3, 6), getFactors(6));
    assertEquals(setOf(1, 2, 4, 8, 16), getFactors(16));
    assertEquals(setOf(1, 2, 3, 4, 6, 12), getFactors(12));
    assertEquals(setOf(1, 2, 4, 5, 10, 20, 25, 50, 100), getFactors(100));
  }


  @Test
  public void lot_of_primes() {
      for (int i = 2; i < 50; i++)
          if (PRIMES_BELOW_50.contains(i))
              assertTrue(isPrime(i));
          else
              assertFalse(isPrime(i));
      for (Integer i : PRIMES_BELOW_50)
          if (i != 2)
              assertFalse(isPrime(i + 1));
          else
            assertTrue(isPrime(i + 1));
  }

  @Test
  public void next_prime() {
    for (int i = 2; i < 48; i++)
      if (! isPrime(i))
        assertTrue(isPrime(PRIMES_BELOW_50.get(binarySearch(PRIMES_BELOW_50, nextPrimeFrom(i)) - 1)));
  }

}

File: scala/primes/src/test/scala/PrimeTest.scala
/**
 * (probably) Copyright 2013 by Neal Ford. All rights reserved.
 */
import org.scalatest.FunSuite

class PrimeTestSuite extends FunSuite {
  def PrimesBelow50 =
    List(2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47)

  test("factors") {
    assert(List(1, 2) == Prime.factors(2))
    assert(List(1, 2, 3, 6) == Prime.factors(6))
    assert(List(1, 2, 4, 8, 16) == Prime.factors(16))
    assert(List(1, 2, 3, 4, 6, 12) == Prime.factors(12))
    assert(List(1, 2, 4, 5, 10, 20, 25, 50, 100) == Prime.factors(100))
  }

  test("primes below 50") {
    for (i <- 2 to 50)
      if (PrimesBelow50.contains(i))
        assert(Prime.isPrime(i) == true)
      else
        assert(Prime.isPrime(i) == false)
    for (i <- PrimesBelow50)
      if (i != 2)
        assert(Prime.isPrime(i+1) == false)
  }
}


